### PR TITLE
HEEDLS-806 return no users for no email and test this

### PR DESCRIPTION
**Pull request by JBsoftwire**
### JIRA link
_[HEEDLS-806](https://softwiretech.atlassian.net/browse/HEEDLS-806)_

### Description
Attempting to get all users with a blank email returns zero users rather than all users with no email set.

### Screenshots
N/A

-----
### Developer checks
(Leave tasks unticked if they haven't been appropriate for your ticket.)

I have:
- [x] Run the formatter and made sure there are no IDE errors.
- [x] Written tests for the changes (accessibility tests, unit tests for controller, data services, services, view models, etc)
- [x] Manually tested my work with and without JavaScript. Full manual testing guidelines can be found here: https://softwiretech.atlassian.net/wiki/spaces/HEE/pages/6703648740/Testing
- [ ] Updated/added documentation in Swiki and/or Readme. Links (if any) are below:
  - [doc_1_here](link_1_here)
- [ ] Updated my Jira ticket with information about other parts of the system that were touched as part of the MR and have to be sanity tested to ensure nothing’s broken.
- [x] Scanned over my own MR to ensure everything is as expected.

### CODE REVIEW

File: DigitalLearningSolutions.Data.Tests/Services/UserServiceTests.cs [hunk 1]
```diff
@@ -1094,6 +1242,22 @@ public void DeactivateOrDeleteAdmin_calls_delete_if_admin_does_not_have_admin_se
             }
         }
 
+        [Test]
+        public void GetUsersByEmailAddress_returns_no_results_for_empty_address()
```
Comment (AlexJacksonDS) [inline DigitalLearningSolutions.Data.Tests/Services/UserServiceTests.cs:1246]: Is it worth adding the 3 test cases null, empty, and whitespace to this test?

File: DigitalLearningSolutions.Data/Services/UserService.cs [hunk 2]
```diff
@@ -348,6 +353,10 @@ CentreAnswersData centreAnswersData
         {
             var delegateUser = userDataService.GetDelegateUserById(editDelegateDetailsData.DelegateId);
             var (adminUser, delegateUsers) = GetUsersByEmailAddress(delegateUser!.EmailAddress!);
+            if (delegateUsers.Count == 0)
+            {
+                delegateUsers.Add(delegateUser);
+            }
```
Comment (AlexJacksonDS) [inline DigitalLearningSolutions.Data/Services/UserService.cs:359]: Is there a particular reason to do this this way, rather than bypassing the UpdateDelegateAccountDetails (line 373) call with no other delegate accounts? UpdateDelegate (line 380) updates the names and email on the single delegate anyway.

**Review by AlexJacksonDS — CHANGES_REQUESTED** [review]

File: DigitalLearningSolutions.Data/Services/UserService.cs [hunk 3]
```diff
@@ -108,6 +108,11 @@ ISessionDataService sessionDataService
 
         public (AdminUser? adminUser, List<DelegateUser> delegateUsers) GetUsersByEmailAddress(string emailAddress)
         {
+            if (string.IsNullOrWhiteSpace(emailAddress))
+            {
+                return (null, new List<DelegateUser>());
+            }
```
Comment (DanBloxham-sw) [inline DigitalLearningSolutions.Data/Services/UserService.cs:114]: Does this impact anywhere else that this method is used?

Additionally, we should probably make the `emailAddress` param nullable as we do handle that case here.

File: DigitalLearningSolutions.Data/Services/UserService.cs [hunk 4]
```diff
@@ -348,6 +353,10 @@ CentreAnswersData centreAnswersData
         {
             var delegateUser = userDataService.GetDelegateUserById(editDelegateDetailsData.DelegateId);
             var (adminUser, delegateUsers) = GetUsersByEmailAddress(delegateUser!.EmailAddress!);
```
Comment (DanBloxham-sw) [inline DigitalLearningSolutions.Data/Services/UserService.cs:355]: The email address here can definitely be null, so worth removing the final `!`

**Review by DanBloxham-sw — CHANGES_REQUESTED** [review]
Some minor things about pre-existing method that has been modified, and I agree with the suggestions in Alex's comments.

File: DigitalLearningSolutions.Data/Services/UserService.cs (hunk 3 above)
Comment (JBsoftwire) [reply DigitalLearningSolutions.Data/Services/UserService.cs:114]: The only potential issue I could find is that `NewEmailAddressIsValid` will no longer report a clash if you try to save an empty email, but empty emails should already have been filtered out by form validation so it's meaningless in practice.

File: DigitalLearningSolutions.Data.Tests/Services/UserServiceTests.cs [hunk 5]
```diff
@@ -1093,61 +1093,6 @@ public void UpdateUserAccountDetailsViaDelegateAccount_calls_UpdateDelegateProfe
             ).MustHaveHappened();
         }
 
-        [Test]
-        public void UpdateUserAccountDetailsViaDelegateAccount_updates_single_account_if_no_email_set()
```
Comment (DanBloxham-sw) [inline DigitalLearningSolutions.Data.Tests/Services/UserServiceTests.cs]: I think we still want this test in, but modify the assertions to check that the call to update multiple delegates only has an empty array of ids passed in.

**Review by DanBloxham-sw — CHANGES_REQUESTED** [review]
Looks good, but it seems a unit test has been removed that I think we should keep and modify slightly.

**Review by DanBloxham-sw — APPROVED** [review]
Yep, think this should cover everything.
